File: src/com/maddyhome/idea/vim/extension/surround/VimSurroundExtension.java
/*
 * IdeaVim - Vim emulator for IDEs based on the IntelliJ platform
 * Copyright (C) 2003-2016 The IdeaVim authors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package com.maddyhome.idea.vim.extension.surround;

import com.google.common.collect.ImmutableMap;
import com.intellij.openapi.actionSystem.DataContext;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.util.Pair;
import com.maddyhome.idea.vim.VimPlugin;
import com.maddyhome.idea.vim.command.CommandState;
import com.maddyhome.idea.vim.command.MappingMode;
import com.maddyhome.idea.vim.command.SelectionType;
import com.maddyhome.idea.vim.common.Mark;
import com.maddyhome.idea.vim.common.TextRange;
import com.maddyhome.idea.vim.extension.VimExtensionHandler;
import com.maddyhome.idea.vim.extension.VimNonDisposableExtension;
import com.maddyhome.idea.vim.group.ChangeGroup;
import com.maddyhome.idea.vim.helper.EditorHelper;
import com.maddyhome.idea.vim.key.OperatorFunction;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import java.awt.event.KeyEvent;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

import static com.maddyhome.idea.vim.extension.VimExtensionFacade.*;
import static com.maddyhome.idea.vim.helper.StringHelper.parseKeys;

/**
 * Port of vim-surround.
 *
 * See https://github.com/tpope/vim-surround
 *
 * @author dhleong
 * @author vlan
 */
public class VimSurroundExtension extends VimNonDisposableExtension {

  private static final char REGISTER = '"';

  private static final Map<Character, Pair<String, String>> SURROUND_PAIRS = ImmutableMap.<Character, Pair<String, String>>builder()
    .put('b', Pair.create("(", ")"))
    .put('(', Pair.create("( ", " )"))
    .put(')', Pair.create("(", ")"))
    .put('B', Pair.create("{", "}"))
    .put('{', Pair.create("{ ", " }"))
    .put('}', Pair.create("{", "}"))
    .put('r', Pair.create("[", "]"))
    .put('[', Pair.create("[ ", " ]"))
    .put(']', Pair.create("[", "]"))
    .put('a', Pair.create("<", ">"))
    .put('>', Pair.create("<", ">"))
    .build();

  @NotNull
  @Override
  public String getName() {
    return "surround";
  }

  @Override
  protected void initOnce() {
    putExtensionHandlerMapping(MappingMode.N, parseKeys("<Plug>YSurround"), new YSurroundHandler(), false);
    putExtensionHandlerMapping(MappingMode.N, parseKeys("<Plug>CSurround"), new CSurroundHandler(), false);
    putExtensionHandlerMapping(MappingMode.N, parseKeys("<Plug>DSurround"), new DSurroundHandler(), false);
    putExtensionHandlerMapping(MappingMode.VO, parseKeys("<Plug>VSurround"), new VSurroundHandler(), false);

    putKeyMapping(MappingMode.N, parseKeys("ys"), parseKeys("<Plug>YSurround"), true);
    putKeyMapping(MappingMode.N, parseKeys("cs"), parseKeys("<Plug>CSurround"), true);
    putKeyMapping(MappingMode.N, parseKeys("ds"), parseKeys("<Plug>DSurround"), true);
    putKeyMapping(MappingMode.VO, parseKeys("S"), parseKeys("<Plug>VSurround"), true);
  }

  @Nullable
  private static Pair<String, String> getSurroundPair(char c) {
    if (SURROUND_PAIRS.containsKey(c)) {
      return SURROUND_PAIRS.get(c);
    }
    else if (!Character.isLetter(c)) {
      final String s = String.valueOf(c);
      return Pair.create(s, s);
    }
    else {
      return null;
    }
  }

  @Nullable
  private static Pair<String, String> inputTagPair(@NotNull Editor editor) {
    final String tagInput = inputString(editor, "<");
    if (tagInput.endsWith(">")) {
      final String tagName = tagInput.substring(0, tagInput.length() - 1);
      return Pair.create("<" + tagName + ">", "</" + tagName + ">");
    }
    else {
      return null;
    }
  }

  @Nullable
  private static Pair<String, String> getOrInputPair(char c, @NotNull Editor editor) {
    return c == '<' || c == 't' ? inputTagPair(editor) : getSurroundPair(c);
  }

  private static char getChar(@NotNull Editor editor) {
    final KeyStroke key = inputKeyStroke(editor);
    final char keyChar = key.getKeyChar();
    if (keyChar == KeyEvent.CHAR_UNDEFINED || keyChar == KeyEvent.VK_ESCAPE) {
      return 0;
    }
    return keyChar;
  }

  private static class YSurroundHandler implements VimExtensionHandler {
    @Override
    public void execute(@NotNull Editor editor, @NotNull DataContext context) {
      setOperatorFunction(new Operator());
      executeNormal(parseKeys("g@"), editor);
    }
  }

  private static class VSurroundHandler implements VimExtensionHandler {
    @Override
    public void execute(@NotNull Editor editor, @NotNull DataContext context) {
      final TextRange visualRange = VimPlugin.getMark().getVisualSelectionMarks(editor);
      if (visualRange == null) {
        return;
      }

      // NB: Operator ignores SelectionType anyway
      if (!new Operator().apply(editor, context, SelectionType.CHARACTER_WISE)) {
        return;
      }

      // Leave visual mode
      executeNormal(parseKeys("<Esc>"), editor);

      editor.getCaretModel().moveToOffset(visualRange.getStartOffset());
    }

  }

  private static class CSurroundHandler implements VimExtensionHandler {
    @Override
    public void execute(@NotNull Editor editor, @NotNull DataContext context) {
      final char charFrom = getChar(editor);
      if (charFrom == 0) {
        return;
      }

      final char charTo = getChar(editor);
      if (charTo == 0) {
        return;
      }

      Pair<String, String> newSurround = getOrInputPair(charTo, editor);
      if (newSurround == null) {
        return;
      }

      change(editor, charFrom, newSurround);
    }

    static void change(@NotNull Editor editor, char charFrom, @Nullable Pair<String, String> newSurround) {
      // We take over the " register, so preserve it
      final List<KeyStroke> oldValue = getRegister(REGISTER);

      // Extract the inner value
      perform("di" + pick(charFrom), editor);
      List<KeyStroke> innerValue = getRegister(REGISTER);
      if (innerValue == null) {
        innerValue = new ArrayList<KeyStroke>();
      }

      // Delete the surrounding
      perform("da" + pick(charFrom), editor);

      // Insert the surrounding characters and paste
      if (newSurround != null) {
        innerValue.addAll(0, parseKeys(escape(newSurround.first)));
        innerValue.addAll(parseKeys(escape(newSurround.second)));
      }
      pasteSurround(innerValue, editor);

      // Restore the old value
      setRegister(REGISTER, oldValue);

      // Jump back to start
      executeNormal(parseKeys("`["), editor);
    }

    @NotNull
    private static String escape(@NotNull String sequence) {
      return sequence.replace("<", "\\<");
    }

    private static void perform(@NotNull String sequence, @NotNull Editor editor) {
      executeNormal(parseKeys("\"" + REGISTER + sequence), editor);
    }

    private static void pasteSurround(@NotNull List<KeyStroke> innerValue, @NotNull Editor editor) {
      // This logic is direct from vim-surround
      final int offset = editor.getCaretModel().getOffset();
      final int lineEndOffset = EditorHelper.getLineEndForOffset(editor, offset);

      final Mark motionEndMark = VimPlugin.getMark().getMark(editor, ']');
      final int motionEndOffset;
      if (motionEndMark != null) {
        motionEndOffset = EditorHelper.getOffset(editor, motionEndMark.getLogicalLine(), motionEndMark.getCol());
      }
      else {
        motionEndOffset = -1;
      }
      final String pasteCommand = motionEndOffset == lineEndOffset && offset + 1 == lineEndOffset ? "p" : "P";
      setRegister(REGISTER, innerValue);
      perform(pasteCommand, editor);
    }

    private static char pick(char charFrom) {
      switch (charFrom) {
        case 'a': return '>';
        case 'r': return ']';
        default: return charFrom;
      }
    }
  }

  private static class DSurroundHandler implements VimExtensionHandler {
    @Override
    public void execute(@NotNull Editor editor, @NotNull DataContext context) {
      // Deleting surround is just changing the surrounding to "nothing"
      final char charFrom = getChar(editor);
      if (charFrom == 0) {
        return;
      }
      CSurroundHandler.change(editor, charFrom, null);
    }
  }

  private static class Operator implements OperatorFunction {
    @Override
    public boolean apply(@NotNull Editor editor, @NotNull DataContext context, @NotNull SelectionType selectionType) {
      final char c = getChar(editor);
      if (c == 0) {
        return true;
      }
      final Pair<String, String> pair = getOrInputPair(c, editor);
      if (pair == null) {
        return false;
      }
      // XXX: Will it work with line-wise or block-wise selections?
      final TextRange range = getSurroundRange(editor);
      if (range == null) {
        return false;
      }
      final ChangeGroup change = VimPlugin.getChange();
      final String leftSurround = pair.getFirst();
      change.insertText(editor, range.getStartOffset(), leftSurround);
      change.insertText(editor, range.getEndOffset() + leftSurround.length(), pair.getSecond());

      // Jump back to start
      executeNormal(parseKeys("`["), editor);
      return true;
    }

    @Nullable
    private TextRange getSurroundRange(@NotNull Editor editor) {
      final CommandState.Mode mode = CommandState.getInstance(editor).getMode();
      switch (mode) {
        case COMMAND:
          return VimPlugin.getMark().getChangeMarks(editor);
        case VISUAL:
          final TextRange visualRange = VimPlugin.getMark().getVisualSelectionMarks(editor);
          if (visualRange == null) return null;
          final int exclusiveEnd = EditorHelper.normalizeOffset(editor, visualRange.getEndOffset() + 1);
          return new TextRange(visualRange.getStartOffset(), exclusiveEnd);
        default:
          return null;
      }
    }
  }
}

File: src/main/java/com/maddyhome/idea/vim/extension/surround/VimSurroundExtension.kt
/*
 * Copyright 2003-2023 The IdeaVim authors
 *
 * Use of this source code is governed by an MIT-style
 * license that can be found in the LICENSE.txt file or at
 * https://opensource.org/licenses/MIT.
 */
package com.maddyhome.idea.vim.extension.surround

import com.intellij.openapi.actionSystem.DataContext
import com.intellij.openapi.application.runWriteAction
import com.intellij.openapi.diagnostic.logger
import com.intellij.openapi.editor.Editor
import com.maddyhome.idea.vim.KeyHandler
import com.maddyhome.idea.vim.VimPlugin
import com.maddyhome.idea.vim.api.ExecutionContext
import com.maddyhome.idea.vim.api.VimCaret
import com.maddyhome.idea.vim.api.VimEditor
import com.maddyhome.idea.vim.api.endsWithNewLine
import com.maddyhome.idea.vim.api.getLeadingCharacterOffset
import com.maddyhome.idea.vim.api.globalOptions
import com.maddyhome.idea.vim.api.injector
import com.maddyhome.idea.vim.api.setChangeMarks
import com.maddyhome.idea.vim.command.MappingMode
import com.maddyhome.idea.vim.command.OperatorArguments
import com.maddyhome.idea.vim.common.TextRange
import com.maddyhome.idea.vim.extension.ExtensionHandler
import com.maddyhome.idea.vim.extension.VimExtension
import com.maddyhome.idea.vim.extension.VimExtensionFacade
import com.maddyhome.idea.vim.extension.VimExtensionFacade.executeNormalWithoutMapping
import com.maddyhome.idea.vim.extension.VimExtensionFacade.getRegisterForCaret
import com.maddyhome.idea.vim.extension.VimExtensionFacade.inputKeyStroke
import com.maddyhome.idea.vim.extension.VimExtensionFacade.inputString
import com.maddyhome.idea.vim.extension.VimExtensionFacade.putExtensionHandlerMapping
import com.maddyhome.idea.vim.extension.VimExtensionFacade.putKeyMappingIfMissing
import com.maddyhome.idea.vim.extension.VimExtensionFacade.setRegisterForCaret
import com.maddyhome.idea.vim.extension.exportOperatorFunction
import com.maddyhome.idea.vim.group.findBlockRange
import com.maddyhome.idea.vim.helper.exitVisualMode
import com.maddyhome.idea.vim.helper.keyStroke
import com.maddyhome.idea.vim.helper.vimKeyStroke
import com.maddyhome.idea.vim.key.OperatorFunction
import com.maddyhome.idea.vim.newapi.ij
import com.maddyhome.idea.vim.newapi.vim
import com.maddyhome.idea.vim.options.helpers.ClipboardOptionHelper
import com.maddyhome.idea.vim.put.PutData
import com.maddyhome.idea.vim.state.mode.Mode
import com.maddyhome.idea.vim.state.mode.SelectionType
import com.maddyhome.idea.vim.state.mode.selectionType
import org.jetbrains.annotations.NonNls
import java.awt.event.KeyEvent
import javax.swing.KeyStroke

/**
 * Port of vim-surround.
 *
 * See https://github.com/tpope/vim-surround
 *
 * @author dhleong
 * @author vlan
 */
internal class VimSurroundExtension : VimExtension {

  override fun getName() = "surround"

  @NonNls
  private val NO_MAPPINGS = "surround_no_mappings"

  override fun init() {
    putExtensionHandlerMapping(MappingMode.N, injector.parser.parseKeys("<Plug>YSurround"), owner, YSurroundHandler(), false)
    putExtensionHandlerMapping(MappingMode.N, injector.parser.parseKeys("<Plug>Yssurround"), owner, YSSurroundHandler(), false)
    putExtensionHandlerMapping(MappingMode.N, injector.parser.parseKeys("<Plug>CSurround"), owner, CSurroundHandler(), false)
    putExtensionHandlerMapping(MappingMode.N, injector.parser.parseKeys("<Plug>DSurround"), owner, DSurroundHandler(), false)
    putExtensionHandlerMapping(MappingMode.XO, injector.parser.parseKeys("<Plug>VSurround"), owner, VSurroundHandler(), false)

    val noMappings = VimPlugin.getVariableService().getGlobalVariableValue(NO_MAPPINGS)?.asBoolean() ?: false
    if (!noMappings) {
      putKeyMappingIfMissing(MappingMode.N, injector.parser.parseKeys("ys"), owner, injector.parser.parseKeys("<Plug>YSurround"), true)
      putKeyMappingIfMissing(MappingMode.N, injector.parser.parseKeys("yss"), owner, injector.parser.parseKeys("<Plug>Yssurround"), true)
      putKeyMappingIfMissing(MappingMode.N, injector.parser.parseKeys("cs"), owner, injector.parser.parseKeys("<Plug>CSurround"), true)
      putKeyMappingIfMissing(MappingMode.N, injector.parser.parseKeys("ds"), owner, injector.parser.parseKeys("<Plug>DSurround"), true)
      putKeyMappingIfMissing(MappingMode.XO, injector.parser.parseKeys("S"), owner, injector.parser.parseKeys("<Plug>VSurround"), true)
    }

    VimExtensionFacade.exportOperatorFunction(OPERATOR_FUNC, Operator())
  }

  private class YSurroundHandler : ExtensionHandler {
    override val isRepeatable = true

    override fun execute(editor: VimEditor, context: ExecutionContext, operatorArguments: OperatorArguments) {
      injector.globalOptions().operatorfunc = OPERATOR_FUNC
      executeNormalWithoutMapping(injector.parser.parseKeys("g@").map { it.keyStroke }, editor.ij)
    }
  }

  private class YSSurroundHandler : ExtensionHandler {
    override val isRepeatable = true

    override fun execute(editor: VimEditor, context: ExecutionContext, operatorArguments: OperatorArguments) {
      val ijEditor = editor.ij
      val c = getChar(ijEditor)
      if (c.code == 0) return
      val pair = getOrInputPair(c, ijEditor, context.ij) ?: return

      editor.forEachCaret {
        val line = it.getBufferPosition().line
        val lineStartOffset = editor.getLeadingCharacterOffset(line)
        val lineEndOffset = editor.getLineEndOffset(line)
        val lastNonWhiteSpaceOffset = getLastNonWhitespaceCharacterOffset(editor.text(), lineStartOffset, lineEndOffset)
        if (lastNonWhiteSpaceOffset != null) {
          val range = TextRange(lineStartOffset, lastNonWhiteSpaceOffset + 1)
          performSurround(pair, range, it)
        }
//        it.moveToOffset(lineStartOffset)
      }
      // Jump back to start
      if (editor.mode !is Mode.NORMAL) {
        editor.mode = Mode.NORMAL()
      }
      executeNormalWithoutMapping(injector.parser.parseKeys("`[").map { it.keyStroke }, ijEditor)
    }

    private fun getLastNonWhitespaceCharacterOffset(chars: CharSequence, startOffset: Int, endOffset: Int): Int? {
      var i = endOffset - 1
      while (i >= startOffset) {
        if (!chars[i].isWhitespace()) return i
        --i
      }
      return null
    }
  }

  private class VSurroundHandler : ExtensionHandler {
    override fun execute(editor: VimEditor, context: ExecutionContext, operatorArguments: OperatorArguments) {
      val selectionStart = editor.ij.caretModel.primaryCaret.selectionStart
      // NB: Operator ignores SelectionType anyway
      if (!Operator().apply(editor, context, editor.mode.selectionType)) {
        return
      }
      runWriteAction {
        // Leave visual mode
        editor.exitVisualMode()
        editor.ij.caretModel.moveToOffset(selectionStart)

        // Reset the key handler so that the command trie is updated for the new mode (Normal)
        // TODO: This should probably be handled by ToHandlerMapping.execute
        KeyHandler.getInstance().reset(editor)
      }
    }
  }

  private class CSurroundHandler : ExtensionHandler {
    override val isRepeatable = true

    override fun execute(editor: VimEditor, context: ExecutionContext, operatorArguments: OperatorArguments) {
      val charFrom = getChar(editor.ij)
      if (charFrom.code == 0) return

      val charTo = getChar(editor.ij)
      if (charTo.code == 0) return

      val newSurround = getOrInputPair(charTo, editor.ij, context.ij) ?: return
      runWriteAction { change(editor, context, charFrom, newSurround) }
    }

    companion object {
      fun change(editor: VimEditor, context: ExecutionContext, charFrom: Char, newSurround: SurroundPair?) {
        // Save old register values for carets
        val surroundings = editor.sortedCarets()
          .map {
            val oldValue: List<KeyStroke>? = getRegisterForCaret(editor, context, REGISTER, it)
            setRegisterForCaret(editor, context, REGISTER, it, null)
            SurroundingInfo(editor, context, it, null, oldValue, false)
          }

        // Delete surrounding's content
        perform("di" + pick(charFrom), editor.ij)

        // Add info about surrounding's inner text and location
        surroundings.forEach {
          // Delete surrounding chars if necessary
          val currentSurrounding = getCurrentSurrounding(it.caret, pick(charFrom))
          if (currentSurrounding != null) {
            it.caret.moveToOffset(currentSurrounding.startOffset)
            injector.application.runWriteAction {
              editor.deleteString(currentSurrounding)
            }
          }

          val registerValue = getRegisterForCaret(editor, context, REGISTER, it.caret)
          val innerValue = if (registerValue.isNullOrEmpty()) emptyList() else registerValue
          it.innerText = innerValue

          // Valid surroundings are only those that:
          // - are validly wrapping with surround characters (i.e. parenthesis, brackets, tags, quotes, etc.);
          // - or have non-empty inner text (e.g. when we are surrounding words: `csw"`)
          if (currentSurrounding != null || innerValue.isNotEmpty()) {
            it.isValidSurrounding = true
          }
        }

        surroundings
          .filter { it.isValidSurrounding } // we do nothing with carets that are not inside the surrounding
          .map { surrounding ->
            val innerValue = injector.parser.toPrintableString(surrounding.innerText?.map { it.vimKeyStroke }!!)
            val text = newSurround?.let {
              val trimmedValue = if (newSurround.shouldTrim) innerValue.trim() else innerValue
              it.first + trimmedValue + it.second
            } ?: innerValue
            val textData = PutData.TextData(null, injector.clipboardManager.dumbCopiedText(text), SelectionType.CHARACTER_WISE)
            val putData = PutData(textData, null, 1, insertTextBeforeCaret = true, rawIndent = true, caretAfterInsertedText = false)

            surrounding.caret to putData
          }.forEach {
            injector.put.putTextForCaret(editor, it.first, context, it.second)
          }

        surroundings.forEach {
          it.restoreRegister()
        }

        executeNormalWithoutMapping(injector.parser.parseKeys("`[").map { it.keyStroke }, editor.ij)
      }

      private fun perform(sequence: String, editor: Editor) {
        ClipboardOptionHelper.IdeaputDisabler()
          .use { executeNormalWithoutMapping(injector.parser.parseKeys("\"" + REGISTER + sequence).map { it.keyStroke }, editor) }
      }

      private fun pick(charFrom: Char) = when (charFrom) {
        'a' -> '>'
        'r' -> ']'
        else -> charFrom
      }

      private fun getCurrentSurrounding(caret: VimCaret, char: Char): TextRange? {
        val editor = caret.editor
        val searchHelper = injector.searchHelper
        return when (char) {
          't' -> searchHelper.findBlockTagRange(editor, caret, 1, true)
          '(', ')', 'b' -> findBlockRange(editor, caret, '(', 1, true)
          '[', ']' -> findBlockRange(editor, caret, '[', 1, true)
          '{', '}', 'B' -> findBlockRange(editor, caret, '{', 1, true)
          '<', '>' -> findBlockRange(editor, caret, '<', 1, true)
          '`', '\'', '"' -> {
            val caretOffset = caret.offset
            val text = editor.text()
            if (text.getOrNull(caretOffset - 1) == char && text.getOrNull(caretOffset) == char) {
              TextRange(caretOffset - 1, caretOffset + 1)
            } else {
              searchHelper.findBlockQuoteInLineRange(editor, caret, char, true)
            }
          }
          'p' -> searchHelper.findParagraphRange(editor, caret, 1, true)
          's' -> searchHelper.findSentenceRange(editor, caret, 1, true)
          else -> null
        }
      }
    }
  }

  private data class SurroundingInfo(
    val editor: VimEditor,
    val context: ExecutionContext,
    val caret: VimCaret,
    var innerText: List<KeyStroke>?,
    val oldRegisterContent: List<KeyStroke>?,
    var isValidSurrounding: Boolean,
  ) {
    fun restoreRegister() {
      setRegisterForCaret(editor, context, REGISTER, caret, oldRegisterContent)
    }
  }

  private class DSurroundHandler : ExtensionHandler {
    override val isRepeatable = true

    override fun execute(editor: VimEditor, context: ExecutionContext, operatorArguments: OperatorArguments) {
      // Deleting surround is just changing the surrounding to "nothing"
      val charFrom = getChar(editor.ij)
      LOG.debug("DSurroundHandler: charFrom = $charFrom")
      if (charFrom.code == 0) return

      runWriteAction { CSurroundHandler.change(editor, context, charFrom, null) }
    }
  }

  private class Operator : OperatorFunction {
    override fun apply(editor: VimEditor, context: ExecutionContext, selectionType: SelectionType?): Boolean {
      val ijEditor = editor.ij
      val c = getChar(ijEditor)
      if (c.code == 0) return true

      val pair = getOrInputPair(c, ijEditor, context.ij) ?: return false
      // XXX: Will it work with line-wise or block-wise selections?
      val range = getSurroundRange(editor.currentCaret()) ?: return false
      performSurround(pair, range, editor.currentCaret(), selectionType == SelectionType.LINE_WISE)
      // Jump back to start
      executeNormalWithoutMapping(injector.parser.parseKeys("`[").map { it.keyStroke }, ijEditor)
      return true
    }

    private fun getSurroundRange(caret: VimCaret): TextRange? {
      val editor = caret.editor
      if (editor.mode is Mode.CMD_LINE) {
        editor.mode = editor.mode.returnTo
      }
      return when (editor.mode) {
        is Mode.NORMAL -> injector.markService.getChangeMarks(caret)
        is Mode.VISUAL -> caret.run { TextRange(selectionStart, selectionEnd) }
        else -> null
      }
    }
  }
}

private val LOG = logger<VimSurroundExtension>()

private const val REGISTER = '"'

private const val OPERATOR_FUNC = "SurroundOperatorFunc"

private val tagNameAndAttributesCapturePattern = "(\\S+)([^>]*)>".toPattern()

private data class SurroundPair(val first: String, val second: String, val shouldTrim: Boolean)

private val SURROUND_PAIRS = mapOf(
  'b' to SurroundPair("(", ")", false),
  '(' to SurroundPair("( ", " )", false),
  ')' to SurroundPair("(", ")", true),
  'B' to SurroundPair("{", "}", false),
  '{' to SurroundPair("{ ", " }", false),
  '}' to SurroundPair("{", "}", true),
  'r' to SurroundPair("[", "]", false),
  '[' to SurroundPair("[ ", " ]", false),
  ']' to SurroundPair("[", "]", true),
  'a' to SurroundPair("<", ">", false),
  '>' to SurroundPair("<", ">", false),
  's' to SurroundPair(" ", "", false),
)

private fun getSurroundPair(c: Char): SurroundPair? = if (c in SURROUND_PAIRS) {
  SURROUND_PAIRS[c]
} else if (!c.isLetter()) {
  val s = c.toString()
  SurroundPair(s, s, false)
} else {
  null
}

private fun inputTagPair(editor: Editor, context: DataContext): SurroundPair? {
  val tagInput = inputString(editor, context, "<", '>')
  if (editor.vim.mode is Mode.CMD_LINE) {
    editor.vim.mode = editor.vim.mode.returnTo
  }
  val matcher = tagNameAndAttributesCapturePattern.matcher(tagInput)
  return if (matcher.find()) {
    val tagName = matcher.group(1)
    val tagAttributes = matcher.group(2)
    SurroundPair("<$tagName$tagAttributes>", "</$tagName>", false)
  } else {
    null
  }
}

private fun inputFunctionName(
  editor: Editor,
  context: DataContext,
  withInternalSpaces: Boolean,
): SurroundPair? {
  val functionNameInput = inputString(editor, context, "function: ", null)
  if (editor.vim.mode is Mode.CMD_LINE) {
    editor.vim.mode = editor.vim.mode.returnTo
  }
  if (functionNameInput.isEmpty()) return null
  return if (withInternalSpaces) {
    SurroundPair("$functionNameInput( ", " )", false)
  } else {
    SurroundPair("$functionNameInput(", ")", false)
  }
}

private fun getOrInputPair(c: Char, editor: Editor, context: DataContext): SurroundPair? = when (c) {
  '<', 't' -> inputTagPair(editor, context)
  'f' -> inputFunctionName(editor, context, false)
  'F' -> inputFunctionName(editor, context, true)
  else -> getSurroundPair(c)
}

private fun getChar(editor: Editor): Char {
  val key = inputKeyStroke(editor)
  val keyChar = key.keyChar
  val res = if (keyChar == KeyEvent.CHAR_UNDEFINED || keyChar.code == KeyEvent.VK_ESCAPE) {
    0.toChar()
  } else {
    keyChar
  }
  LOG.trace("getChar: $res")
  return res
}

private fun performSurround(pair: SurroundPair, range: TextRange, caret: VimCaret, tagsOnNewLines: Boolean = false) {
  runWriteAction {
    val editor = caret.editor
    val change = VimPlugin.getChange()
    val leftSurround = pair.first + if (tagsOnNewLines) "\n" else ""

    val isEOF = range.endOffset == editor.text().length
    val hasNewLine = editor.endsWithNewLine()
    val rightSurround = if (tagsOnNewLines) {
      if (isEOF && !hasNewLine) {
        "\n" + pair.second
      } else {
        pair.second + "\n"
      }
    } else {
      pair.second
    }

    change.insertText(editor, caret, range.startOffset, leftSurround)
    change.insertText(editor, caret, range.endOffset + leftSurround.length, rightSurround)
    injector.markService.setChangeMarks(
      caret,
      TextRange(range.startOffset, range.endOffset + leftSurround.length + rightSurround.length)
    )
  }
}
